src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+
+
+  // Copy assignment
+  RawImageEigen& RawImageEigen::operator=(const RawImageEigen& source) {
+    width = source.width;
+    height = source.height;
+    image = source.image;
+    obstime = source.obstime;
+    return *this;
+  }
+
+
+  // Move assignment
+  RawImageEigen& RawImageEigen::operator=(RawImageEigen&& source) {
+    if (this != &source) {
+      width = source.width;
+      height = source.height;
+      image = std::move(source.image);
+      obstime = source.obstime;
+    }
+    return *this;
+  }
+
+
+  bool RawImageEigen::isclose(const RawImageEigen& img_b, float atol) const {
+    return image.isApprox(img_b.image, atol);
+  }
+
+
+  auto RawImageEigen::get_interp_neighbors_and_weights(const float x, const float y) const {
+    Point p(x, y);
+
+    // coordinates of nearest neighboring array
+    // top left (tl), top right (tr),
+    // bottom left (bl), bottom right (br)
+    auto [tl, tr, bl, br] = p.nearest_pixel_coords();
+
+    // interpolation weights for each of the 4 neighbors
+    // total normalization should be 1 always (size of array)
+    // float normalization = 1 / ((tr.x - tl.x) * (tl.y - bl.y));
+    float w_tl = (tr.x - x) * (y - bl.y);
+    float w_tr = (x - tl.x) * (y - bl.y);
+    float w_bl = (br.x - x) * (tr.y - y);
+    float w_br = (x - bl.x) * (tl.y - y);
+
+    auto [itl, itr, ibl, ibr] = p.nearest_pixel_idxs();
+
+    struct retval {
+      // gee golly naming is hard...
+      Index top_left, top_right, bot_left, bot_right;
+      float w_tl, w_tr, w_bl, w_br;
+    };
+
+    return retval{itl, itr, ibl, ibr, w_tl, w_tr, w_bl, w_br};
+  }
+
+
+  float RawImageEigen::interpolate2(const float x, const float y) const {
+    Point p(x, y);
+
+    // neighbors and weights for interpolation
+    auto iwn = get_interp_neighbors_and_weights(x, y);
+
+    // I sure hope the compiler optimizes locals away successfully
+    auto tl = iwn.top_left;
+    auto tr = iwn.top_right;
+    auto bl = iwn.bot_left;
+    auto br = iwn.bot_right;
+
+    // dodge masked values (could also have been an expression,
+    // or for loop, not sure about preferred C++ ways or benefits?
+    //return (w_tl * (arr(tl.x, tl.y) == NO_DATA) ? 0.0f, arr(tl.x, tl.y) +
+    //        w_tr * (arr(tr.x, tr.y) == NO_DATA) ? 0.0f, arr(tl.x, tl.y) +
+    //        w_bl * (arr(bl.x, bl.y) == NO_DATA) ? 0.0f, arr(tl.x, tl.y) +
+    //        w_br * (arr(br.x, br.y) == NO_DATA) ? 0.0f, arr(tl.x, tl.y));
+    float interpolated_value = 0.0f;
+    if (image(tl.i, tl.j) != NO_DATA)
+      interpolated_value += iwn.w_tl * image(tl.i, tl.j);
+
+    if (image(tr.i, tr.j) != NO_DATA)
+      interpolated_value += iwn.w_tr * image(tr.i, tr.j);
+
+    if (image(bl.i, bl.j) != NO_DATA)
+      interpolated_value += iwn.w_bl * image(bl.i, bl.j);
+
+    if (image(br.i, br.j) != NO_DATA)
+      interpolated_value += iwn.w_br * image(br.i, br.j);
+
+    if (interpolated_value == 0.0f)
+      return NO_DATA;
+    return interpolated_value;
```

Comment:
I think this is different in than the result of the original code. In the original code the interpolated value was the weighted average of the pixels without `NO_DATA`. So if a pixel was `NO_DATA` you did not include it's value in the sum or its weight in the averaging function. Handles the case of excluding the value, but not that case of excluding the weight. So `NO_DATA` pixels will be treated as 0.0.

Reply:
Fixed.